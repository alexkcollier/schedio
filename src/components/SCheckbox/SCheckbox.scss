.checkbox {
  @include check-or-radio {
    border-radius: border-radius('small');
  }

  & ~ &__check::after {
    background-color: transparent !important;
    border: border-width('thick') solid color('green', 'light');
    border-left: none;
    border-top: none;
    content: '';
    height: 0.65em;
    left: 50%;
    opacity: 0;
    position: absolute;
    top: 45%;
    transform: translateX(-50%) translateY(-50%) rotate(45deg);
    transition: all duration() ease;
    width: 0.35em;
  }

  &:checked ~ &__check::after {
    opacity: 1;
  }
}
