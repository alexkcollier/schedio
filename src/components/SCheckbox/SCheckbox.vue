<template>
  <SLabel
    :for="id"
    :label="label"
    :is-reversed="isReversed"
    :is-required="required"
    :hide-optional="hideOptional"
    is-inline
  >
    <input
      :id="id"
      v-bind="$attrs"
      :name="computedName"
      :required="required"
      class="checkbox"
      type="checkbox"
      :checked="!!value"
      v-on="listeners"
    >
    <span
      :class="{ 'checkbox__check--invalid': isInvalid }"
      class="checkbox__check"
    />
  </SLabel>
</template>

<script>
import InputControl from '@/mixins/InputControl.mixin'

export default {
  name: 'SCheckbox',

  mixins: [InputControl],

  props: {
    value: {
      type: null,
      default: null,
    },
  },
}
</script>

<style lang="scss">
.checkbox {
  @include check-or-radio {
    border-radius: border-radius('small');
  }

  & ~ &__check::after {
    background-color: transparent !important;
    border: border-width('thick') solid color('green', 'light');
    border-left: none;
    border-top: none;
    content: '';
    height: 0.65rem;
    left: 50%;
    opacity: 0;
    position: absolute;
    top: 45%;
    transform: translateX(-50%) translateY(-50%) rotate(45deg);
    transition: all duration() ease;
    width: 0.35rem;
  }

  &:checked ~ &__check::after {
    opacity: 1;
  }
}
</style>
