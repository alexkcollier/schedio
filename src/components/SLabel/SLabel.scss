.label {
  @include font-size(5);
  display: block;
  font-weight: $font-weight-bold;
  text-transform: uppercase;

  &--inline {
    align-items: center;
    display: inline-flex;
    flex-direction: row;
    justify-content: flex-start;

    .input,
    .textarea,
    .select {
      margin-left: spacing();
    }

    &.label--reversed {
      flex-direction: row-reverse;
    }
  }

  &--fullwidth {
    width: 100%;
  }

  &:not(:last-child) {
    margin-right: spacing();
  }

  &:not(:first-child) {
    margin: spacing('loose') 0;
  }

  &__optional {
    @include font-size(6);
    color: color('night', 'lightest');
    font-weight: $font-weight-light;
    text-transform: none;
  }
}
