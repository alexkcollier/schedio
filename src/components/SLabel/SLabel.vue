<template>
  <label :class="classList" class="label">
    <!-- Wrapped to keep optional on the right when reversed -->
    <span>
      {{ label }}
      <span v-if="isOptional" class="label__optional">&mdash;&nbsp;Optional</span>
    </span>
    <slot/>
  </label>
</template>

<script>
export default {
  name: 'SLabel',

  props: {
    label: {
      type: String,
      required: true
    },

    isOptional: {
      type: Boolean,
      default: true
    },

    isInline: {
      type: Boolean,
      default: false
    },

    isFullwidth: {
      type: Boolean,
      default: false
    },

    isReversed: {
      type: Boolean,
      default: false
    }
  },

  computed: {
    classList() {
      return {
        'label--inline': this.isInline,
        'label--fullwidth': this.isFullwidth,
        'label--reversed': this.isReversed
      }
    }
  },

  mounted() {
    if ((this.isFullwidth || this.isReversed) && !this.isInline) {
      console.error('`isFullwidth` and `isReversed` require `isInline`')
    }
  }
}
</script>
