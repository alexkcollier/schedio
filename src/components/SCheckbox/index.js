import SCheckbox from './SCheckbox.vue'

export { SCheckbox }

export default { install: Vue => Vue.component(SCheckbox.name, SCheckbox) }
