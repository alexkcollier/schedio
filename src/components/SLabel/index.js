import SLabel from './SLabel.vue'

export { SLabel }

export default { install: Vue => Vue.component(SLabel.name, SLabel) }
