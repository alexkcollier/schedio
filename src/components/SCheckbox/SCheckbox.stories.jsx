import { SCheckbox } from '@/components/SCheckbox';
import { SHeading } from '@/components/SHeading';
import { withAttrsAsProps, withUnboundAttrs } from '@/mixins/stories/form-fields';
import PropList from '@@/docs/components/PropList';
import { boolean } from '@storybook/addon-knobs';

export default {
  title: 'Components/Inputs/SCheckbox',

  parameters: {
    jest: 'SCheckbox',
  },
};

export const checkbox = () => ({
  props: {
    props: {
      default: () => ({
        ...withAttrsAsProps({
          id: 'base-checkbox',
          name: 'base-checkbox',
          label: 'Base Checkbox',
        }),
        isReversed: boolean('is-reversed', false, 'Optional Props'),
      }),
    },
    attrs: { default: withUnboundAttrs({ value: 'base-checkbox' }) },
  },

  render (h) {
    const { attrs, props } = this.$props;
    return (
      <div>
        <SHeading level="1">Checkbox</SHeading>
        <p>
          Checkboxes provide the ability to make a range of selections (none, one, or multiple).
          They can also be used to provide consent, or indicate agreement.
        </p>

        <SHeading level="2">Example</SHeading>
        <SCheckbox {...{ props }} {...{ attrs }} />

        <PropList component={SCheckbox} />
      </div>
    );
  },
});

checkbox.story = {
  name: 'Checkbox',
};
