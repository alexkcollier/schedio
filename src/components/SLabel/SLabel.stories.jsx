import { SHeading } from '@/components/SHeading'
import { SLabel } from '@/components/SLabel'
import PropList from '@@/docs/components/PropList'
import { boolean, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'

storiesOf('Components/Inputs.SLabel', module)
  .addParameters({ jest: 'SLabel' })
  .add('Label', () => {
    return {
      props: {
        props: {
          default: () => ({
            label: text('Label text', 'Label text', 'Required Props'),
            isFullwidth: boolean('isFullwidth', false, 'Optional Props'),
            isInline: boolean('isInline', false, 'Optional Props'),
            isReversed: boolean('isReversed', false, 'Optional Props'),
            isRequired: boolean('iRrequired', false, 'Optional Props'),
            hideOptional: boolean('hide-optional', false, 'Optional Props')
          })
        },

        defaultSlot: {
          default: text('Default', 'SomeInputComponent', 'Slots')
        }
      },
      render(h) {
        const { props, defaultSlot } = this.$props
        return (
          <div>
            <SHeading level="1">Label</SHeading>
            <p>
              <code>SLabel</code> provides labels for all form inputs. It is included automatically
              for all input components in this library, and is designed to wrap components.
            </p>
            <p>
              By default, if a label is for an optional field, it will append '- Optional' to the
              label. This text can be hidden with the <code>hide-optional</code> prop.
            </p>

            <SHeading level="2">Example</SHeading>
            <SLabel {...{ props }}>
              <div style="text-transform: none; font-weight: 300;">
                <code>{'<' + defaultSlot + ' />'}</code>
              </div>
            </SLabel>

            <PropList component={SLabel} />
          </div>
        )
      }
    }
  })
