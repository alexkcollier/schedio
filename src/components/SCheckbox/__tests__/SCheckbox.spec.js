import { shallowMount } from '@vue/test-utils'
import SCheckbox from '@/components/SCheckbox/SCheckbox.vue'
import SLabel from '@/components/SLabel/SLabel.vue'

describe('SCheckbox.vue', () => {
  const inputEvent = jest.fn()
  let wrapper
  let checkbox

  beforeEach(() => {
    wrapper = shallowMount(SCheckbox, {
      propsData: {
        id: 'checkbox',
        name: 'checkbox',
        label: 'Checkbox'
      },
      listeners: {
        input: inputEvent
      }
    })

    checkbox = wrapper.find('#checkbox')
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('requires required props', () => {
    // spies on `console.error` without calling `console.error`
    const spy = jest.spyOn(global.console, 'error').mockImplementation(() => {})
    // mounting component without required props
    shallowMount(SCheckbox)
    expect(spy).toBeCalled()
    expect(spy.mock.calls[0][0]).toContain('[Vue warn]: Missing required prop')
  })

  it('can be checked and unchecked', () => {
    // check
    wrapper.trigger('click')
    expect(checkbox.element.checked).toBe(true)

    // uncheck
    wrapper.trigger('click')
    expect(checkbox.element.checked).toBe(false)

    // check that events were emitted correctly
    expect(wrapper.emitted('input')).toEqual([[true], [false]])
  })

  it('can be a required input', () => {
    wrapper.setProps({ required: true })
    expect(checkbox.element.required).toBe(true)
  })

  it('can be invalid', () => {
    wrapper.setProps({ isInvalid: true })
    expect(checkbox.element.validity.valid).toBe(false)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can be disabled', () => {
    const disabledWrapper = shallowMount(SCheckbox, {
      propsData: {
        id: 'checkbox',
        name: 'checkbox',
        label: 'Checkbox'
      },
      attrs: {
        disabled: true
      }
    })

    // verifies disabled attr is passed through to checkbox input element
    expect(disabledWrapper.attributes('disabled')).toBeUndefined()
    expect(disabledWrapper.find('#checkbox').element.disabled).toBeTruthy()
  })

  it('can be reversed', () => {
    wrapper.setProps({ isReversed: true })
    expect(wrapper.find(SLabel).props('isReversed')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
