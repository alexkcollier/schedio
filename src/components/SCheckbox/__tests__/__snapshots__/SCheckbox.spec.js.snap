// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SCheckbox.vue can be invalid 1`] = `
<label class="label label--inline" for="checkbox"><span>
    Checkbox
    <!----></span> <input id="checkbox" name="checkbox" required="required" type="checkbox" class="checkbox"> <span class="checkbox__check checkbox__check--invalid"></span></label>
`;

exports[`SCheckbox.vue can be reversed 1`] = `
<label class="label label--inline label--reversed" for="checkbox"><span>
    Checkbox
    <!----></span> <input id="checkbox" name="checkbox" required="required" type="checkbox" class="checkbox"> <span class="checkbox__check"></span></label>
`;

exports[`SCheckbox.vue can hide the optional tag 1`] = `
<label class="label label--inline" for="checkbox"><span>
    Checkbox
    <!----></span> <input id="checkbox" name="checkbox" required="required" type="checkbox" class="checkbox"> <span class="checkbox__check"></span></label>
`;

exports[`SCheckbox.vue renders correctly 1`] = `
<label class="label label--inline" for="checkbox"><span>
    Checkbox
    <!----></span> <input id="checkbox" name="checkbox" required="required" type="checkbox" class="checkbox"> <span class="checkbox__check"></span></label>
`;
